File: src/main/scala/com/vc/easy/L206.scala
package com.vc.easy

class ListNode206(var _x: Int = 0) {
  var next: ListNode206 = null
  var x: Int = _x
}

object L206 {
  def reverseList(head: ListNode206): ListNode206 = {
    if(head == null) return head
    else {
      val dummyNode = new ListNode206()
      dummyNode.next = head

      val prev = dummyNode
      val current = prev.next
      var next = current.next

      //0->1->2->3->4->5->NULL
      //p->c->n->3->4->5->NULL

      //0->2->1->3->4->5->NULL
      //0->3->2->1->4->5->NULL
      //0->4->3->2->1->5->NULL
      //0->5->4->3->2->1->NULL
      while(next != null) {
        current.next = next.next
        next.next = prev.next
        prev.next = next
        next = current.next
      }
      dummyNode.next
    }
  }
}

File: src/main/scala/com/vc/easy/L206.java
package com.vc.easy;

class ListNode206 {
    int val;
    ListNode206 next;
    ListNode206(int x) { val = x; }
}

class Solution {
    public ListNode206 reverseList(ListNode206 head) {
        if(head == null) return null;

        ListNode206 prev = new ListNode206(-1);
        prev.next = head;
        ListNode206 current = prev.next;
        ListNode206 next = current.next;
        while(next != null) {
            current.next = next.next;
            next.next = prev.next;
            prev.next = next;

            /**
                 p     c    n
                 d =>  1 -> 2 -> 3 -> 4 -> 5

                 p     n    c
                 d =>  2 -> 1 -> 3 -> 4 -> 5

                 p          c    n
                 d =>  2 -> 1 -> 3 -> 4 -> 5
             */
            next = current.next;
        }
        return prev.next;
    }
}
